### ISSUE TITLE
Is heap command support qemu-user ?

### ISSUE THREAD

**Issue by lonnywong** [pwndbg/pwndbg#822]
### Description

```
/tmp/pwndbg/aarch64_struct-malloc_state.cc:200:21: error: aggregate ‘malloc_state foo’ has incomplete type and cannot be defined
  200 | struct malloc_state foo;
      |                     ^~~
Traceback (most recent call last):
  File "/root/pwndbg/pwndbg/commands/__init__.py", line 130, in __call__
    return self.function(*args, **kwargs)
  File "/root/pwndbg/pwndbg/commands/__init__.py", line 221, in _OnlyWhenRunning
    return function(*a, **kw)
  File "/root/pwndbg/pwndbg/commands/__init__.py", line 259, in _OnlyWithLibcDebugSyms
    return function(*a, **kw)
  File "/root/pwndbg/pwndbg/commands/__init__.py", line 239, in _OnlyWhenHeapIsInitialized
    return function(*a, **kw)
  File "/root/pwndbg/pwndbg/commands/heap.py", line 100, in heap
    heap_region = allocator.get_heap_boundaries(addr)
  File "/root/pwndbg/pwndbg/heap/ptmalloc.py", line 357, in get_heap_boundaries
    region = self.get_region(addr) if addr else self.get_region(self.get_arena()['top'])
  File "/root/pwndbg/pwndbg/heap/ptmalloc.py", line 329, in get_arena
    return self.main_arena
  File "/root/pwndbg/pwndbg/heap/ptmalloc.py", line 72, in main_arena
    self._main_arena = pwndbg.memory.poi(self.malloc_state, main_arena_addr)
  File "/root/pwndbg/pwndbg/memoize.py", line 44, in __call__
    value = self.func(*args, **kwargs)
  File "/root/pwndbg/pwndbg/heap/ptmalloc.py", line 190, in malloc_state
    return pwndbg.typeinfo.load('struct malloc_state')
  File "/root/pwndbg/pwndbg/typeinfo.py", line 157, in load
    return gdb.lookup_type(name)
gdb.error: No struct type named malloc_state.
```

### Steps to reproduce
```sh
apt-get update
apt-get install qemu-user
apt-get install libc6-*arm*-cross
wget https://github.com/bkerler/exploit_me/raw/master/bin/exploit64
chmod +x exploit64
qemu-aarch64 -g 1234 -L /usr/aarch64-linux-gnu exploit64 Exploiter A
```
```sh
apt-get install gdb-multiarch
gdb-multiarch exploit64
pwndbg> target remote localhost:1234
pwndbg> b *0x400f10
pwndbg> c
pwndbg> heap
```

### My setup

```
Gdb:      9.1
Python:   3.8.2 (default, Jul 16 2020, 14:00:26)  [GCC 9.3.0]
Pwndbg:   1.1.0 build: 487caa1
Capstone: 4.0.1024
Unicorn:  1.0.1
```

**Comment by disconnect3d** [comment]
Hey,

qemu-user is under-developed due to the lack of us getting vmmap information.

However... I think the heap integration should maybe work now? @lebr0nli WDYT?

**Comment by lebr0nli** [comment]
> However... I think the heap integration should maybe work now? @lebr0nli WDYT?

Ahh, unfortunately, for this binary, not yet 😢
Here are some things we should enhance in the future (if possible haha):


1. For the main cause of the issue above, it's that although we can find `main_arena` symbols, we cannot find `struct malloc_state`. So maybe we should try to access some symbols directly first even under heap heuristic or use some heuristics on type loading even under debug symbols mode.
2. The method for finding the GLIBC version should be updated. Is it possible to detect the GLIBC version without searching `"GNU C Library XXXXXXX"` in memory when there is no `__libc_version` symbol? Because this file doesn't have the symbol and that kind of string. Or maybe we should have an interface to let the user manually set it.
3. The assembly code of some libc functions in the static library of this binary seem has some differences from the assembly code in the dynamic library I tested before. I can still try to add support to it, but the implementation of it will be very dirty I guess... So maybe I should use a more advanced technique to reverse the assembly code instead of looking the result of capstone only in the future. @disconnect3d , do you have any experience with reverse engineering about this? Or do you have any suggestions about which Python module I can try? Ideally, if we can do something like ida pro's `F5` does, it will be easier for finding the symbol's reference inside the function. 
4. Somehow, the heuristic of GEF we used for `main_arena` failed again lol. Maybe we should always use our heuristic when we have more confidence in it.

----

To summarize, IMO if we want to fix this issue for now, 1 + 2 are enough, but if we want to support all situations in the future, we will need 3 + 4 maybe.
So for now, I'll try something about 1 + 2 first to fix this issue.

**Comment by gsingh93** [comment]
The primary issue here should be fixed, the heap command should work on qemu-user.
